File: backend_fastapi/app/routers/voice.py
from __future__ import annotations

import base64

from fastapi import APIRouter
from pydantic import BaseModel

from ..prompts import render_prompt
from ..tts import synthesize_tts_wav

router = APIRouter(prefix="/api/voice", tags=["voice"])


class GeneratePromptRequest(BaseModel):
    scenario: str
    language: str


class GeneratePromptResponse(BaseModel):
    success: bool
    systemPrompt: str


@router.post("/generate-prompt", response_model=GeneratePromptResponse)
async def generate_prompt(req: GeneratePromptRequest) -> GeneratePromptResponse:
    scenario = (req.scenario or "").strip()
    language = (req.language or "").strip()

    # 最小可用：不依赖 LLM 可用性，直接用模板生成 system prompt。
    # 前端会允许用户在 review step 里编辑。
    system_prompt = render_prompt("voice_system_prompt.j2", scenario=scenario, language=language)

    return GeneratePromptResponse(success=True, systemPrompt=system_prompt)


class StartSessionRequest(BaseModel):
    systemPrompt: str


class StartSessionResponse(BaseModel):
    success: bool
    openingText: str
    openingAudio: str


@router.post("/start", response_model=StartSessionResponse)
async def start_session(req: StartSessionRequest) -> StartSessionResponse:
    system_prompt = (req.systemPrompt or "").strip()

    # 为了兼容现有前端流程，返回固定开场白；同时生成可播放的 openingAudio。
    # openingAudio: base64(wav bytes)。默认 TTS 后端是 silence（仍是“真实音频”，只是静音）。
    if system_prompt:
        opening_text = "好的，我们开始练习吧。你可以先说一句话。"
    else:
        opening_text = "我们开始吧。你可以先说一句话。"

    wav = synthesize_tts_wav(opening_text)
    opening_audio_b64 = base64.b64encode(wav).decode("utf-8") if wav else ""
    return StartSessionResponse(success=True, openingText=opening_text, openingAudio=opening_audio_b64)

File: app/v5/src/services/voice.ts
/**
 * @fileoverview 语音业务服务模块 (Voice Service)
 * @description 提供语音对话相关的 HTTP 接口调用，主要用于会话初始化和 Prompt 生成。
 *              注意：实时的语音流传输通过 WebSocket (voice-socket.ts) 处理，而非本模块。
 */

import api from './api';

/** 语音 Prompt 生成结果接口 */
export interface VoicePromptResult {
  success: boolean;
  /** 生成的系统提示词，用于指导 AI 的角色扮演 */
  systemPrompt: string;
}

/** 语音会话启动结果接口 */
export interface VoiceStartResult {
  success: boolean;
  /** AI 的开场白文本 */
  openingText: string;
  /** AI 开场白的音频数据（base64(wav bytes)） */
  openingAudio: string; 
}

export const VoiceService = {
  /**
   * 生成系统提示词 (System Prompt)
   * 
   * 根据用户选择的场景和语言，调用后端生成对应的 AI 角色设定提示词。
   * 
   * @param {string} scenario - 对话场景 ID (如 'daily', 'business')
   * @param {string} language - 目标语言代码 (如 'zh-CN', 'en-US')
   * @returns {Promise<string>} 生成的系统提示词
   * @throws {Error} 如果生成失败
   */
  async generatePrompt(scenario: string, language: string): Promise<string> {
    const response = await api.post('/api/voice/generate-prompt', { scenario, language });
    const result = response as unknown as VoicePromptResult;
    if (result.success) {
      return result.systemPrompt;
    }
    throw new Error('生成提示词失败');
  },

  /**
   * 启动语音会话
   * 
   * 向后端发送系统提示词，初始化一个新的语音会话上下文。
   * 后端会返回 AI 的开场白（文本和音频）。
   * 
   * @param {string} systemPrompt - 系统提示词
   * @returns {Promise<VoiceStartResult>} 会话启动结果
   * @throws {Error} 如果启动失败
   */
  async startSession(systemPrompt: string): Promise<VoiceStartResult> {
    const response = await api.post('/api/voice/start', { systemPrompt });
    const result = response as unknown as VoiceStartResult;
    if (result.success) {
      return result;
    }
    throw new Error('启动会话失败');
  }
};
